modules/gdscript/gdscript_parser.cpp

```diff
@@ -4929,10 +4964,42 @@ void GDScriptParser::_parse_class(ClassNode *p_class) {
 							member._export.hint_string = member.data_type.native_type;
 							member._export.class_name = member.data_type.native_type;
 						} else {
-							_set_error("Invalid export type. Only built-in and native resource types can be exported.", member.line);
+							_set_error("Invalid native export type. \"" + member.data_type.native_type + "\" is not a Resource type.", member.line);
+							return;
+						}
+					} else if (member.data_type.kind == DataType::SCRIPT || member.data_type.kind == DataType::GDSCRIPT) {
+						if (member.data_type.script_type.is_null()) {
+							_set_error("The member's data type is a script that did not load successfully.", member.line);
+							return;
+						}
+						StringName class_name = ScriptServer::get_global_class_name(member.data_type.script_type->get_path());
+						if (class_name == StringName()) {
+							_set_error("Invalid script export type. \"" + class_name + "\" is not a script class.", member.line);
+							return;
+						}
+						if (ClassDB::is_parent_class(member.data_type.native_type, "Resource")) {
+							member._export.type = Variant::OBJECT;
+							member._export.hint = PROPERTY_HINT_RESOURCE_TYPE;
+							member._export.usage |= PROPERTY_USAGE_SCRIPT_VARIABLE;
+							member._export.hint_string = class_name;
+							member._export.class_name = class_name;
+						} else {
+							_set_error("Invalid script class export type. \"" + class_name + "\" is not a Resource type.", member.line);
 							return;
 						}
 
+					} else if (member.data_type.kind == DataType::UNRESOLVED && ScriptServer::is_global_class(member.data_type.native_type)) {
+						String class_name = member.data_type.native_type;
+						if (ScriptServer::get_global_class_native_base(class_name) == "Resource") {
```

Comment:
I have some confusion about these lines.

1. If `member.data_type.native_type` is a StringName containing the native class type, then why would it ever become a script class name? I'm not sure of what the intent here is. I was looking in the source for where `native_type` gets assigned a value, but I couldn't seem to find anything in `gdscript_parser.h/.cpp`, so I'm not sure how a script class name is supposed to end up there. I did see that `parse_class_name()` directly assigns the parsed identifier to `current_class->identifier`, so that is likely what you would want to pass to the ScriptServer, unless I'm mistaken.

1. Based on the above point, I think you should gather the `class_name` variable's value from something different. As far as I can tell with how it is used, `native_type` is supposed to be an actual C++ engine class.

1. I don't think you can simply check if the native type of the class is "Resource". After all, someone might choose to extend a different class other than Resource and make a script class out of *that* type. So, you should first grab the native type, and then compare it to the ClassDB in full.

        if (ClassDB::is_parent_class(native_type, "Resource")) { ... }

Reply:
1. In 3.2 stable the parser sets the `native_type` to the identifier's string here:
https://github.com/godotengine/godot/blob/3.2/modules/gdscript/gdscript_parser.cpp#L5723-L5731

I agree it doesn't make much sense semantically, but I just took the information I had available, I didn't change any of the type parsing code.

3. Totally agree, I will update the PR soon.